Add experimental `ResponsiveBlockControl` component

getdave opened the pull request:
# PLEASE READ THIS BEFORE REVIEWING 😀

This PR isn't in any way wired up to Block state. Therefore changing the values in the UI doesn't do anything or persist any settings whatsoever. 

This is purely **a PR for the UI only** to allow multiple controls for different viewports only. It is a sub PR of a wider effort to allow the Group Block to define spacing controls (eg: margin / padding). See below for more details.

# Description

Sub PR of [Adds basic dimension controls to Group Block](https://github.com/WordPress/gutenberg/pull/16730).

Adds a standardised `<ResponsiveBlockControl>` component which provides a UI for Block controls that need to have responsive settings. 

Features:

- Standardised, accessible markup using `fieldset` and `legend` 
- Complete control over the rendered controls via render props
- Render prop to render a default control (non-responsive mode)
- Toggle to turn responsive controls on/off
- Render prop to render a Responsive control(s) (responsive mode)



## How has this been tested?

I've temporarily hacked and implementation into the `core/group` Block which utilises `<select>` inputs and temporary `state` for demo purposes. 

Note that **this will be removed before merging**.

## Screenshots

#### Default

<img width="280" alt="Screen Shot 2019-07-29 at 11 48 09" src="https://user-images.githubusercontent.com/444434/62042780-cc598200-b1f6-11e9-8345-6dd41baaf946.png">

#### Responsive (toggled)
<img width="283" alt="Screen Shot 2019-07-29 at 11 48 18" src="https://user-images.githubusercontent.com/444434/62042779-cc598200-b1f6-11e9-896e-eca1ef49f2b5.png">



## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Basic component skeleton

getdave committed:
Adds component demo to Group Block

getdave committed:
Make toggle control label dynamic and translatable

getdave committed:
Adds styles

karmatosed commented:
Adding a design review here there are a few points come to mind:

![image](https://user-images.githubusercontent.com/253067/62365981-7c8ff900-b51d-11e9-95ba-2bdb16b26977.png)

- It struck me why no large option. I found that super interesting to experience. I think that's covered in another PR though #16791 maybe?
- I have to admit it was hard to guess what small/medium would be.
- I can see people wanting to set the spacing themselves. This is where I think we might be getting into a slippery slope of mammoth configurations but I am noting it.
- I wonder if using term mobile is future proof? You use your iPad not just mobile.

shaunandrews commented:
Took some time to work up a design.

<img width="680" alt="image" src="https://user-images.githubusercontent.com/191598/62392875-9d187d00-b536-11e9-9530-090825a4d176.png">

One thing you'll notice is I've reverse the default state of the Switch to be on by default. Switching it off would show the multiple rows.

An alternative ideas is to list out all available options with a Switch for each row. This tends to work great if we plan to have more than 2 options, but otherwise feels a little busy and overwhelming.

<img width="690" alt="image" src="https://user-images.githubusercontent.com/191598/62393455-3d22d600-b538-11e9-9768-e278a797c901.png">

--

I mentioned some related changes to the placeholder value over in #16791

shaunandrews commented:
Another consideration here is that its likely we'll have many responsive settings across blocks and even within settings for a single block. For example, consider the need to adjust not only basic spacing (like we have above) but also font-size, or color, or general layout. At some point we may end up with a "mobile" switch for each of these settings. This is going to be problematic.

What we might want to do instead, is have some sort of Switch for the entire editor. Something along the lines of a "mobile mode" or a "tablet mode" that would allow changes to be made to that specific view, without affecting the "desktop mode" display. I realize this is a much larger change than you initial proposed, but I think its the better way of handling the UI.

getdave commented:
> Took some time to work up a design.

Thanks for much for taking the time here @shaunandrews. It is much appreciated.

> I've reverse the default state of the Switch to be on by default. Switching it off would show the multiple rows.

I see why you've done this. Playing devil's advocate here for a second. Does it feel a little odd to be turning something "off" to enable a feature? 

> ...list out all available options with a Switch for each row. This tends to work great if we plan to have more than 2 options, but otherwise feels a little busy and overwhelming.

Nice exploration. I agree this doesn't scale well though.

> At some point we may end up with a "mobile" switch for each of these settings. This is going to be problematic.

Yes that could get a little...noisy UI-wise!

> What we might want to do instead, is have some sort of Switch for the entire editor. Something along the lines of a "mobile mode" or a "tablet mode" that would allow changes to be made to that specific view, without affecting the "desktop mode" display. 

A good idea. I wonder if @youknowriad is aware of any efforts in this area?

> I realize this is a much larger change than you initial proposed, but I think its the better way of handling the UI.

I agree although as you've identified it's a lot more complex. @youknowriad @mtias are you able to advise as to whether you feel the direction we're proposing here is suitable or will it be wasted effort in the long run if you have a "mobile mode" in the works?

mapk commented:
> provides a UI for Block controls that need to have responsive settings.

This seems very related to the responsive block controls that went through several iterations here: https://github.com/WordPress/gutenberg/issues/13363

I think we came super close with this particular design from @kjellr : https://github.com/WordPress/gutenberg/issues/13363#issuecomment-476714596

I understand the UI in #13363 is primarily for columns, but can help inform the design behind spacing too hopefully.

getdave commented:
@mapk Thanks very much for your feedback.

> I think we came super close with this particular design from @kjellr : #13363 (comment)

@mapk This is what I was building on having touched base with @kjellr earlier. However there have been a few iterations since and @shaunandrews has provided some more input. 

Are you able to advise how you feel about the designs @shaunandrews has proposed (cc @kjellr )? I'd like to avoid costly refactors in code because it takes a long time. Ideally, I've love ot get consensus from the design before I write any more code here. 

Much appreciated.

mapk commented:
> Are you able to advise how you feel about the designs @shaunandrews has proposed

For sure! Sorry to have let this stall. Regarding @shaunandrews' mockups, I definitely prefer the first concept, so I'll comment on that one specifically.

![Screen Shot 2019-08-14 at 8 29 54 AM](https://user-images.githubusercontent.com/617986/63034210-b80bbb00-be6d-11e9-9eb1-94c08bc61eb1.png)


1. Looking at this, I first thought the dropdown "Small" was a reference to device size due to the label. To curtail this confusion, we repeated a lot of words in the [other mockups for Columns](https://github.com/WordPress/gutenberg/issues/13363#issuecomment-464853461).
2. I think there will be an a11y problem with switching out the label of the dropdown when toggling on. This will probably not work well with screen-readers.
3. That second point leads me to think if we move the toggle to the top, and keep the dropdowns below it, that might work better. I'm sure some language would have to be changed to support this, but when toggling the top control, it feels okay to have the bottom controls reflect the change.

getdave commented:
Note this a11y comment

https://github.com/WordPress/gutenberg/pull/16791#issuecomment-531163401

getdave committed:
Automatically render responsive controls based on default control

Continue to allow full customisation of the responsive controls, but by default render the responsive fields using the same markup as the default control. This avoids duplication when consuming the component.

getdave committed:
Wrap each field group in a `fieldset` element

getdave committed:
Invert toggle and language to be on by default.

Addresses points raised in https://github.com/WordPress/gutenberg/pull/16790/#issuecomment-517814542

getdave committed:
Adds initial tests

afercia commented:
Quoting from https://github.com/WordPress/gutenberg/pull/16791#issuecomment-531163401To my understanding, this control is meant to add padding / margin settings to the group block, with the ability to specify this setting for different types of devices. If this is the case, I'd suggest to use `<fieldset>` and `<legend>` elements to group the settings based on devices. Pseudo code:
```
fieldset
    visible legend: All devices
    DimensionControl: Margin
    DimensionControl: Padding
/fieldset
fieldset
    visible legend: Small screen devices
    DimensionControl: Margin
    DimensionControl: Padding
/fieldset
etc.
```

getdave committed:
Update tests to resemble how a user interacts with Component

afercia commented:
@getdave I see only now the screenshot and questions before my previous comment 🤦‍♂ 

Yes, in your example the nested fieldset + legend aren't ideal. That would require to change the HTML structure and the text used for legend / labels.

As general rules:
- the fieldset legend gives a name to the group of controls
- each control should have a meaningful label 
- see also https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/

In the example above I'm not sure the labels for the select are ideal.
- Label = "Desktop/Tablet/Mobile" doesn't tell what the values within the select are about
- semantically, that actually means: "Select a small/medium Desktop" which is confusing
- the selects are meant to set the padding values: they contain padding values therefore they should be labelled "Padding"
- To clarify what this "padding" relates to, the fieldset legend would give context by clarifying it's about "Desktop"
- This makes even more sense when there are multiple selects within the same group (e.g. padding, margin)

> Would you say it's overkill to have a fieldset when there is only a single field 

It really depends on the text used for the labels.
In that case, the visible label should also clarify what the "padding" is about, e.g.:
`<label for...>Padding for desktop devices</label>`

It is OK to remove _all_ the fieldsets if each label provides all the necessary information.

> When the toggle is activated how would you recommend changing the focus to the new fieldsets

Couple of considerations:
- I'm not sure a switch toggle is the most appropriate control: switch toggles should be used to set a setting that gets immediately saved without the need of submitting. In this case, it justs switches the visibility of a UI section. I'd consider to use a different control in the first place. /Cc @mapk 
- I'd move the toggle control (whatever it is) _before_ all the selects. This way, focus can stay on the toggle and the UI _after_ that changes. Users can explore the content after the toggle as they would do normally. Instead, if the UI that changes is placed _before_ the toggle, they won't have a clue something has changed.

getdave committed:
Add toggle state test

getdave committed:
Update to switch toggle to preceed controls

Addresses concern raised in https://github.com/WordPress/gutenberg/pull/16790/#issuecomment-531777090

getdave committed:
Update individual control labels to fully describe control for a11y

Address concerns raised in https://github.com/WordPress/gutenberg/pull/16790/#issuecomment-531777090

getdave commented:
@afercia Thanks for your feedback.

Here's a screenshot

<img width="276" alt="Screen Shot 2019-09-17 at 15 50 28" src="https://user-images.githubusercontent.com/444434/65074480-ec89ff80-d962-11e9-8e6e-9e6464e6d904.png">


> Yes, in your example the nested fieldset + legend aren't ideal. That would require to change the HTML structure and the text used for legend / labels.

I agree. So I've removed the `fieldsets` but kept it open for developers to utilise these when it is most appropriate. The rendering of the control itself if fully in the control of the developer so they can choose to use `fieldsets` where that makes sense.

> It really depends on the text used for the labels.
In that case, the visible label should also clarify what the "padding" is about, e.g.:

I believe the label per device approach is better in this case. As a result I've amended the label text to include a full description as you advise. However, previously design feedback is that including the full text is too visually "verbose". Therefore I suggest a compromise of providing the full text to screen readers whilst providing a shortened version for sighted users. I think with the context of the "Padding" fieldset legend surrounding all the controls it is clear enough that the device names refer to the setting for "padding" for that device.

paaljoachim commented:
I feel that the latest mockup showing the controls in the Spacing area is nice and clean.
It is clear what it does. Having an option to use the same padding on all screen sizes (two words?) is nice. Desktop - Tablet - Mobile drop downs are easy to understand.
Thanks for working on this Dave! To me this feels like the right direction and would be a very nice first iteration!

getdave commented:
> I'd also love to see more alignment with previous explorations for responsive settings. This should help with standardization of controls.

I'm going to defer to @mapk on this. I think for sanity's sake (ie: not having to navigate the Issue/PR "backscroll" it would be great to have a single updated design for me to work from that ticks all the boxes you refer to. Once I have this I'll restart work here.

getdave commented:
I think we could also do with technical feedback here @youknowriad.

Particularly, have I made the component flexible enough but still retained sensible defaults? I don't want to make the consumer of this component have to work too hard to implement a reliable and accessible control, but I need to leave the door open for more bespoke use cases.

Bear in mind that we're _not_ going to persist any of the changes to the Group Block. The ones in this PR are just to make it "testable" for folks helping out on this PR.

mapk commented:
I haven't gotten to any design critique yet, but just noticed this functionality issue. The `small` or `medium` settings isn't persistent.

![padding](https://user-images.githubusercontent.com/617986/65608019-6113fe00-df62-11e9-85a3-709838113b52.gif)

mapk commented:
I've noticed a spacing issue (marked in red):

The spacing between the label and the form element need to match. Maybe try aligning the toggle with the top of the text "Use the same padding..."?

![Screen Shot 2019-09-25 at 3 50 39 PM](https://user-images.githubusercontent.com/617986/65645529-b70d9380-dfac-11e9-885b-6363709c84d0.png)

mapk commented:
> I'm going to defer to @mapk on this. I think for sanity's sake (ie: not having to navigate the Issue/PR "backscroll") it would be great to have a single updated design for me to work from that ticks all the boxes you refer to. Once I have this I'll restart work here.

After some testing, I don't think a redesign of the controls is necessary. You have standard dimensions that don't have minor in-between steps. So "small" "medium" etc. make sense to just keep in a dropdown. 

The only two things I've noticed that need to get updated are the spacing above the toggle, and the fact that my selections are not persistent. The selection isn't reflected in Gutenberg or on the frontend either.

getdave commented:
@mapk Thanks v much for your review. 

> I don't think a redesign of the controls is necessary

Great news. I say we run with this simple approach.

> ... the spacing above the toggle

I'll fix that shortly.

> the fact that my selections are not persistent. The selection isn't reflected in Gutenberg or on the frontend either.

That's correct. As it stands this PR is purely to do with the component, how it looks and how it is implemented technically to ensure it's as flexible as possible. It doesn't actually _doi_ much in Gutenberg yet cos it's a component you're expected to consume and wire up to a piece of UI. I think the best explanation is [the one in the PR description](https://github.com/WordPress/gutenberg/pull/16790#issue-302013652) (don't worry it's easy to miss and most people have!). The reason I've done this is because I was asked to break up the parent PR which is _ultimately_ going to use this component to adding spacing controls to the Group Block.

getdave committed:
Fixes form ui to improve alignment

Addresses https://github.com/WordPress/gutenberg/pull/16790#issuecomment-535255247

getdave commented:
Todo

* Update to leave DOM intact when the toggle state changes and use HTML attrs and CSS to manipulate the presence of the controls.
* Add tests for toggle behaviour
* Add tests to cover customising responsive controls via render prop

koke commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,72 @@
+	return (
+		<Fragment>
+			<span className="screen-reader-text">
+				{ property } for
```
I don't think that's going to work for translations. We probably have to build the whole string for screen readers with `sprintf`, and expose it to screen readers in a different way (maybe `aria-label`?).

getdave replied:
Thanks! I thought I'd caught all these on my pass i18n pass, but due to this change to the component it's slipped through again.

I'll get this fixed as well.

getdave replied:
Fixed

koke commented on packages/block-library/src/group/edit.js:
```diff
@@ -43,6 +68,26 @@ function GroupEdit( {
+							} );
+						} }
+						renderDefaultControl={ ( label ) => (
+							<SelectControl
```
Unless there's some other magic taking care of it, I'm missing an `onChange` here. Is this actually saving the settings per-device?

koke replied:
> This isn't in any way wired up to Block state...

I should have read the full description first, sorry 🤦‍♂

getdave replied:
Please see PR description.

getdave committed:
Improves i18n of generated control label

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r330595669.

Note that aria-describedby is the correct type of aria role for this use case https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute

getdave committed:
Improve a11y by leaving DOM intact between state changes.

getdave committed:
Adds aria description to toggle control for improved a11y context

getdave committed:
Update tests

getdave committed:
Update toggle of responsive controls internal to the component

getdave committed:
Adds test for responsiveControlsActive setting prop

getdave committed:
Add tests to cover custom labels and custom device sets

getdave committed:
Fix to only build default repsonsive controls when necessary

getdave committed:
Adds tests to cover rendering of custom responsive controls

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+}
+
+function ResponsiveBlockControl( props ) {
+	const { legend, property, toggleLabel, responsiveControlsActive = false, renderDefaultControl, defaultLabel = __( 'All' ), devices = [ __( 'Desktop' ), __( 'Tablet' ), __( 'Mobile' ) ], renderResponsiveControls } = props;
```
I think the devices need an identifier like a slug that we have for colors. Right now they are only identified with a label that is tralatable we need a unique way to refer a device that does not change with the translation.

getdave replied:
Great point. I will sort this.

getdave replied:
Ok I've converted `devices` into an array of objects with the form:

```js
{
    id: 'small', // unique non-translated
    label: __('Small'), // purely for visual labels
}
```

Ditto for the default value of `All`.

This ensures we always have an unchanging value for the device.

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+	const defaultResponsiveControls = () => {
+		return devices.map( ( deviceLabel, index ) => (
+			<Fragment key={ index }>
+				{ renderDefaultControl( deviceLabel ) }
```
We only pass the label to the renderDefaultControl, in a real case if we were storing the value somewhere e.g: attribute we would need to know an identifier of the size we are referring to. So, besides passing a label I think we need to pass something like a device id.

getdave replied:
Makes sense. I now pass the entire `device` object to the callback to ensure it has access.

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+					onChange={ handleToggle }
+					help={ toggleHelpText }
+				/>
+
```
Why hiding instead of rendering one option or the other depending on isResponsiveMode?

getdave replied:
I was thinking about a11y. I was thinking if you were focused on one set of controls and then you toggled to the other "mode" (eg: "Default" -> "Responsive"). I felt it might be a bit confusing or jarring to have the original controls completely removed from the DOM and a load of new controls suddenly appear.

What do you think @afercia?

jorgefilipecosta replied:
>  I felt it might be a bit confusing or jarring to have the original controls completely removed from the DOM and a load of new controls suddenly appear.

If we leave things on the dom, screen reader users may still be able to interact with the controls? If that's the case, they will interact with controls that make nothing that seems worst.
We have at least another case exactly like this the hasParallax prop in the Cover block. That case does not renders on the dom so I guess we should follow the same approach here. If we find something better from the a11y point of view in this PR I think as a follow-up we should update the other PR.

getdave replied:
Ok noted. Will change.

getdave replied:
Fixed.

jorgefilipecosta commented on packages/block-library/src/group/edit.js:
```diff
@@ -43,6 +66,21 @@ function GroupEdit( {
+						responsiveControlsActive={ true }
+						renderDefaultControl={ ( label ) => (
+							<SelectControl
+								label={ <ResponsiveBlockControlLabel property="Padding" device={ label } /> }
```
Can't the label passed as prop already include the component ResponsiveBlockControlLabel rendered, so we don't need to expose ResponsiveBlockControlLabel at all?

getdave replied:
I've been wrestling with this one.

I've just realised I could pass the `<ResponsiveBlockControlLabel />` component as an argument to the `renderDefaultControl` render prop. This allow the render prop function to consume the label as a component and utilise that however it needs.

In the example case of a `<SelectControl />` we simply pass the label component as the `label` prop and everything "just works". A more complex `renderDefaultControl` use case could use the component directly in a more custom manner.

Thanks for prompting me to reconsider my approach.

Note I've left `ResponsiveBlockControlLabel` exposed as the moment but I guess we could remove the `export`?

getdave replied:
This has now all been resolved by handling the label internally and removing the additional `export`.

jorgefilipecosta reviewed (commented):
Hi @getdave nice work here, things well to me I just left some comments/questions.
I also think we should mark the components we are exposing as __experimental, we are in a starting phase and I would prefer to not lockin a specific API yet allowing us to change things more easily.

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+function ResponsiveBlockControl( props ) {
+	const { legend, property, toggleLabel, responsiveControlsActive = false, renderDefaultControl, defaultLabel = __( 'All' ), devices = [ __( 'Desktop' ), __( 'Tablet' ), __( 'Mobile' ) ], renderResponsiveControls } = props;
+
+	const [ isResponsiveMode, setIsResponsiveMode ] = useState( responsiveControlsActive );
```
We keeping the responsive mode as a local state, but when the flag is toggled the parent needs to know about this state change, otherwise how is the change going to be persisted?

I think the parent should pass something like isResponsiveMode, and an onIsResponsiveMode change and the parent should be responsible for handling and storing changes.

getdave replied:
I've also been wrestling with this one.

You'll note until recently this was controlled by the consuming "parent" component. Then in [this commit, I moved it to become component/local state](https://github.com/WordPress/gutenberg/pull/16790/commits/1a0966b95d8469747640dc9579cae58634229873). 

This was because it felt odd that the consumer needed to provide a means to toggle the UI. I felt the state of the toggle should be internal (thereby avoiding the need for the user to have to know about this implementation detail) but we should expose a callback to notify when it changes and also a means to set the default state.

Note that currently, I've tried to find the best of both worlds. By default you don't need to handle the toggle manually when utilising the component - it is handled internally and "just works" (this is what I'd expect to happen as a user). However, you can optionally set the boolean `responsiveControlsActive` prop to determine the initial state of the UI (ie: open/closed responsive controls). This allows us to persist an open/closed state between renders and/or editor initialisations. I probably need a decent test case for this scenario.

It's simple enough to roll this back to my original implementation (which is what I think you are recommending) but I am concerned about making this component too hard to utilise.

getdave replied:
I've revised this a little. Let me know what you think.

jorgefilipecosta replied:
> This was because it felt odd that the consumer needed to provide a means to toggle the UI.

If the consumer passes isResponsiveMode and onIsResponsiveModeChange the consumer is not passing a means to toggle the UI it is passing if the current mode is responsive and a handler to change it. It is not even aware that the mode is changed by using a toggle.
If responsiveControlsActive changes, the component will not reflect that change, this seems problematic is it makes it impossible to apply a change in the flag from a third party plugin if the component is rendered.

getdave replied:
To be clear

`responsiveControlsActive` - this sets the initial state of the toggle. This allows the consuming component to set the initial mode to responsive or not.

`onIsResponsiveModeChange` - this is a callback purely to "notify" the consuming component that the responsive mode has changed to the value provided as an argument (ie: on/off). It doesn't do any _setting_.

In addition, there is also the _internal_ state value `isResponsiveMode`. This is used to manage the state of the responsive toggle internally. The consuming component has no means to alter this value except on the initial render where the initial value of the state is set by `responsiveControlsActive` prop.

My thinking was that it was too much to ask the consuming component to manage the internal toggle state of the UI. I thought that by managing internally it would make the component easier to use. I recognised that the consuming component would need to be able to set the initial toggle state which is why I went for the `responsiveControlsActive` prop.

That said, recent experience has led me to believe that almost all components that involve user input need to be fully controlled components. I think this is what [you were driving at](https://github.com/WordPress/gutenberg/pull/16790#discussion_r331606578) and I'm coming around to that idea, even if it makes the component more labourious to consume.

getdave commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+		);
+	};
+
+	const handleToggle = ( isChecked ) => {
```
Can I check why you are suggesting we use useCallback here? My understanding that it was effectively to "memoize" computationally expensive functions and/or to ensure referential equality.

I assume you're coming at this with a perf lense? If so do you feel that the function is sufficiently intensive to warrant a `useCallback`?

If I'm missed something I'd be open to an explanation - I'm always looking to improve my understanding, especially in the field of hooks!

jorgefilipecosta replied:
Hi @getdave I answered this question on https://github.com/WordPress/gutenberg/pull/16791#discussion_r332690665.

getdave replied:
Really helpful. Much appreciated. Will remember this one (I hope!).

getdave commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+
+export function ResponsiveBlockControlLabel( { property, device } ) {
+	const accessibleLabel = sprintf( __( 'Controls the %s property for %s devices.' ), property, device );
+	const descId = uniqueId();
```
Thank you for reminding me. I should know about this by now as I've come across that HOC often enough. Apologies.

getdave replied:
Fixed.

getdave committed:
Update to make label component part of component’s internal API

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r331602172.

Also updates tests.

getdave committed:
Adds callback prop to fire when responsive mode state changes

Addresses https://github.com/jorgefilipecosta

getdave committed:
Update to utilise withInstanceId HOC

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r331589219

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,85 @@
+} from '@wordpress/components';
+
+export function ResponsiveBlockControlLabel( { property, device } ) {
+	const accessibleLabel = sprintf( __( 'Controls the %s property for %s devices.' ), property, device );
```
I forgot to tell we also need to number the strings so the comment can refer them %1$s,  %2$s.

getdave replied:
Fixed. Thanks

jorgefilipecosta commented on packages/block-library/src/group/edit.js:
```diff
@@ -43,6 +65,19 @@ function GroupEdit( {
 				/>
+
+				<PanelBody title={ __( 'Spacing' ) }>
+					<ResponsiveBlockControl
```
I guess we should mark the component as experimental for some time.

getdave replied:
Definitely agreed! Done.

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,92 @@
+	ToggleControl,
+} from '@wordpress/components';
+
+export const ResponsiveBlockControlLabel = withInstanceId( function ResponsiveBlockControlLabel( { instanceId, property, device } ) {
```
I think we could remove this export.

getdave replied:
Agreed. Fixed.

RodrigoDAgostino commented:
Seems I'm a bit late to the discussion, but I still would like to ask: wouldn't it be better to have the control be content-oriented instead of device-oriented?
Designing based on device sizes is a bad practice that should be avoided. I think _small_, _medium_ and _large_ would be a better approach than _mobile_, _tablet_ and _desktop_.

getdave commented:
> Seems I'm a bit late to the discussion, but I still would like to ask: wouldn't it be better to have the control be content-oriented instead of device-oriented?
> Designing based on device sizes is a bad practice that should be avoided. I think _small_, _medium_ and _large_ would be a better approach than _mobile_, _tablet_ and _desktop_.

I completely understand and really appreciate you raising this. You are right to do so.

I'm also conscious that we shouldn't be trying to target _devices_.

That said, I'm not sure users of the Block Editor think like developers. They just want to "make it work on mobile, iPad...etc". This issue of perception would be my only concern otherwise I'm more than happy to change this.

What do you think @mapk?

getdave committed:
Removes unused export

This component is now provided by internal API and doesn’t need exposing.

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r332718859

getdave committed:
Mark as “experimental”

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r332716937

getdave committed:
Remove non exposed component API doc

getdave committed:
Extracts label to be a dedicated component

getdave committed:
Updates to completely switch out DOM when responsive mode is toggled

Please see https://github.com/WordPress/gutenberg/pull/16790#discussion_r331599556

getdave committed:
Update to useCallback to avoid expensive DOM re-renders

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r331586167

See also https://github.com/WordPress/gutenberg/pull/16791#discussion_r331624709

getdave committed:
Updates i18n to provide better descriptions for translators

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r331587069

rodrigodagostino commented:
> 
> 
> > Seems I'm a bit late to the discussion, but I still would like to ask: wouldn't it be better to have the control be content-oriented instead of device-oriented?
> > Designing based on device sizes is a bad practice that should be avoided. I think _small_, _medium_ and _large_ would be a better approach than _mobile_, _tablet_ and _desktop_.
> 
> I completely understand and really appreciate you raising this. You are right to do so.
> 
> I'm also conscious that we shouldn't be trying to target _devices_.
> 
> That said, I'm not sure users of the Block Editor think like developers. They just want to "make it work on mobile, iPad...etc". This issue of perception would be my only concern otherwise I'm more than happy to change this.
> 
> What do you think @mapk?

Not long after I published my previous comment, I actually arrived to the same conclusion as you, Dave :P I'm honestly sorry I forgot to share it with you, guys. Some kind of middle-ground between those approaches would be something like _small devices_, _medium devices_, and _large devices_ (which actually is the terminology that I ended up using in my own responsive controls implementation).

I believe something like that would better serve both users and developers :)

getdave committed:
Updates devices list to contain unique non-translatable slug

This addresses concern that we need to be able to identify a “size” by a value that does not change due to translations.

See https://github.com/WordPress/gutenberg/pull/16790#discussion_r331597378

getdave committed:
Ensure renderDefault controls render prop has access to device details

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r331598820

getdave commented:
> Some kind of middle-ground between those approaches would be something like small devices, medium devices, and large devices (which actually is the terminology that I ended up using in my own responsive controls implementation).

No worries @rodrigodagostino. Looks like we're on the same track. I've updated the implementation to follow your approach which could work.

That said, I'd still value input from @mapk and @jorgefilipecosta on this to sanity check us.

getdave committed:
Begin documentation

Incomplete.

jorgefilipecosta commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,112 @@
+		return null;
+	}
+
+	const toggleControlLabel = toggleLabel || sprintf( _x( 'Use the same %s on all screensizes.', 'Toggle control label. Should the property be the same across all screen sizes or unique per screen size.' ), property );
```
We need a translator comment.
An example of a comment is `/* translators: accessibility text. %1: current block position (number). %2: next block position (number) */`

getdave replied:
Got it. I totally misunderstood this before.

I've resolved this, but I have noticed that I cannot see this type of comment mentioned anywhere in the `docs`. Did I miss something or do we need to add this?

jorgefilipecosta reviewed (approved):
I left some comments (some are optional and are just my thoughts), but In general, I think the PR is ready 👍

getdave committed:
Rename “legend” prop to “title”

This prop was named based on implementation details. Switching out for more agnostic term.

getdave committed:
Fix incorrect usage of _x and replace with translator comments

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r340622289

Noting that this type of comment seems to be undocumented in Gutenberg…

getdave committed:
Update internal nomenclature from “device” to “viewport”

This is a small terminology change which could have a big change on the way developers think about repsonsive settings. We shouldn’t be thinking about “devices” but rather “viewports” / screen sizes. We can still present to the user as “devices” but the developer should not be tying layout changes to specific devices (or conceptual groups of devices - eg: “Mobile”).

getdave committed:
Refactor to make component fully controlled

Addresses https://github.com/WordPress/gutenberg/pull/16790#discussion_r340589849 by making the state of the responsive mode controlled by the consuming component. This completes the process of making the component fully controlled.

getdave committed:
Adds custom hook useResponsiveBlockControl

Attempt to relieve some of the overhead associated with having ResponsiveBlockControl be a fully controlled component. By consuming this hook, a developer can wire up a default handler for toggling responsive mode without having to worry about creating their own useState-based hooks.

getdave commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,112 @@
+
+	useEffect( () => {
+		if ( isFunction( onIsResponsiveModeChange ) ) {
+			onIsResponsiveModeChange( isResponsiveMode );
```
No longer required. I've made component fully controlled.

getdave committed:
Revert "Adds custom hook useResponsiveBlockControl"

This reverts commit 48529f70a4f3d17ce2b5d1be57e8d08780f680bc.

getdave committed:
Remove testing implementation in Group Block

getdave committed:
Docs update

rodrigodagostino commented:
@getdave thanks a lot for your efforts and also for considering my opinion :) I think _small screens_, _medium screens_, and _large screens_ is even better than the middle-ground I suggested.
Just wanted to share with you one more little thing: the UI of the spacing panel I've implemented a while ago in my custom blocks.

![Opera Snapshot_2019-11-01_152649_localhost](https://user-images.githubusercontent.com/8670793/68047348-efe00d00-fcbc-11e9-9d28-76e3adac1df5.png)
I know it doesn't really fit the current Gutenberg workflow, but it might be a little helpful in the future.

aduth commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,97 @@
+				/>
+
+				{ ! isResponsive && (
+					<div className="block-editor-responsive-block-control__group block-editor-responsive-block-control__group--default" >
```
Same note here about BEM convention as at: https://github.com/WordPress/gutenberg/pull/17846#discussion_r368140643

Also:

- These classes are only referenced in tests. It's not strictly problematic to do so, but ideally the runtime implementation needn't be conformed specifically to tailor to tests.
- I think in representing this as some sort of on/off state (it either _is_ or _isn't_ "responsive"), we'd not need to refer to this explicitly as "default" but rather as the absence of "responsive". In other words, I'd suggest using modifier class `is-responsive` for the other condition, and for this one, simply omitting it.

The second point would actually enable us to simplify this implementation, where currently there is code duplication for `block-editor-responsive-block-control__group` largely because of how this class is assigned. With the above changes, it could become a unified wrapper of:

```jsx
<div className={ classnames( 'block-editor-responsive-block-control__group', { 'is-responsive': isResponsive } ) } hidden={ ! isResponsive }>
```

aduth commented on packages/block-editor/src/components/responsive-block-control/index.js:
```diff
@@ -0,0 +1,97 @@
+				) }
+
+				{ isResponsive && (
+					<div className="block-editor-responsive-block-control__group block-editor-responsive-block-control__group--responsive" hidden={ ! isResponsive }>
```
The assignment of `hidden` here is effectively dead code, because this `div` is only rendered inside the condition of `isResponsive && <div />`.

aduth commented:
Regarding https://github.com/WordPress/gutenberg/pull/16790#discussion_r368149717 and https://github.com/WordPress/gutenberg/pull/16790#discussion_r368149922 , I have proposed relevant changes at #19738.

tellthemachines commented on packages/block-editor/src/components/responsive-block-control/test/__snapshots__/index.js.snap:
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Any idea how this snapshot ended up minified? Makes it impossible to review changes to the component 😞

aduth replied:
@tellthemachines It's quite likely that because the test is [using a _string_ of HTML](https://github.com/WordPress/gutenberg/blob/11eaaee2e3cfdb62c40b7d131c7bd021da1fca99/packages/block-editor/src/components/responsive-block-control/test/index.js#L121) for the snapshot, it doesn't know how to make sense of the string as markup of elements to break across lines, vs. say a React element tree. I'd agree though that it sort of defeats the purpose of a snapshot if a future maintainer is unable to assess whether changes in the snapshot are sensible.

I'm not sure exactly what our current configuration will support for those snapshots, but I'd wonder if it could be using the DOM element or if we could adapt somehow for the React element to be used.
